Assignment6.java

```diff
@@ -77,24 +77,28 @@
      * @param indexMap the map from cars to slots
      * @return next car to be moved (0 in case of rearranged array)
      */
-    private static int nextCarToMove(int[] given, int[] desired, Map<Integer, Integer> indexMap) {
+    private static int nextCarToMove(int[] given, int[] desired, Map<Integer, Integer> indexMap, HashSet<Integer> misplacedCars) {
         //arrays of length 0
         if (given.length == 0) {
             return 0;
         }
         
         //car belonging to the current empty slot
         if (desired[indexMap.get(0)] != 0) {
+            // car will be moved in place next, remove it from set of misplaced cars
+            misplacedCars.remove(desired[indexMap.get(0)]);
             return desired[indexMap.get(0)];
+
         }
-        
+
         //empty lot in its desired place => deblock the algorithm
-        //find the first car that isn't in its right place
-        for (int i = 0; i < given.length; i++) {
-            if (given[i] != 0 && given[i] != desired[i]) {
-                return given[i];
-            }
+        // pick a random car from set of misplaced cars

```

Comment:
Because I think it leads to the same number of moves (I think someone else commented that on slack too, and I agree). At least on the test examples I quickly ran it on. Did you find an example where we observe a better behavior?